quantum/quantum_read_write_mutex.h

```diff
@@ -93,32 +102,48 @@ class ReadWriteMutex
     /// @brief Returns the number of readers holding the lock.
     /// @return The number of readers.
     int numReaders() const;
+    
+    /// @brief Returns the number of pending upgraded writers.
+    /// @return The number of writers.
+    int numPendingWriters() const;
 
-    class ReadGuard
+    class Guard
     {
     public:
         /// @brief Construct this object and lock the passed-in mutex as a reader.
         /// @param[in] lock ReadWriteMutex which protects a scope during the lifetime of the Guard.
+        /// @param[in] acquire Determines the type of ownership.
         /// @note Blocks the current thread until the mutex is acquired.
-        explicit ReadGuard(ReadWriteMutex& lock);
-        ReadGuard(ICoroSync::Ptr sync,
-                  ReadWriteMutex& lock);
+        Guard(ReadWriteMutex& lock,
+              LockTraits::AcquireRead acquire);
+        Guard(ReadWriteMutex& lock,
+              LockTraits::AcquireWrite acquire);
+        Guard(ICoroSync::Ptr sync,
+              ReadWriteMutex& lock,
+              LockTraits::AcquireRead acquire);
+        Guard(ICoroSync::Ptr sync,
+              ReadWriteMutex& lock,
+              LockTraits::AcquireWrite acquire);
         
         /// @brief Construct this object and tries to lock the passed-in mutex as a reader.
         /// @param[in] lock ReadWriteMutex which protects a scope during the lifetime of the Guard.
+        /// @param[in] acquire Determines the type of ownership.
         /// @note Attempts to lock the mutex. Does not block.
-        ReadGuard(ReadWriteMutex& lock,
-                  ReadWriteMutex::TryToLock);
+        Guard(ReadWriteMutex& lock,
+              LockTraits::AcquireRead acquire,
+              LockTraits::TryToLock);
+        Guard(ReadWriteMutex& lock,
+              LockTraits::AcquireWrite acquire,
+              LockTraits::TryToLock);
         
-        /// @brief Construct this object and does not lock the mutex. Assumes the application already
-        ///        owns the read lock.
+        /// @brief Construct this object and assumes the current state of the lock w/o modifying it.
         /// @param[in] lock ReadWriteMutex which protects a scope during the lifetime of the Guard.
-        /// @note Attempts to lock the mutex. Does not block.
-        ReadGuard(ReadWriteMutex& lock,
-                  ReadWriteMutex::AdoptLock);
+        /// @note Does not block.
+        Guard(ReadWriteMutex& lock,
+              LockTraits::AdoptLock);
         
         /// @brief Destroy this object and unlock the underlying mutex if this object owns it.
```

Comment:
Rephrase this as "unlock the underlying mutex if this object still owns a lock state on it" or something similar. That is, the object doesn't really give up ownership of the Mutex object (unless we implement release() again). It always owns the Mutex, what it may or may not own is the locked state.